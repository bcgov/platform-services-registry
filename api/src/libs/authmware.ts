//
// SecureImage
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-02-14.
//

/* eslint-env es6 */

'use strict';

import { getJwtCertificate, logger } from '@bcgov/common-nodejs-utils';
import passport from 'passport';
import { ExtractJwt, Strategy as JwtStrategy } from 'passport-jwt';
import config from '../config';
import { WEB_CLIENT_ID } from '../constants';
import DataManager from '../db';
import shared from './shared';

export interface AuthenticatedUser {
  id: number;
  givenName: string;
  familyName: string;
  name: string;
  preferredUsername: string;
  email: string;
  archived: boolean;
  roles: string[];
  lastSeenAt: object,
}

export const isAuthorized = jwtPayload => {
  return true;
};

export const getJwtPayloadRoles = (jwtPayload: object): string[] | [] => {
  try {
    /* tslint:disable-next-line */
    return jwtPayload['resource_access'][WEB_CLIENT_ID]['roles'];
  } catch (err) {
    return [];
  }
};

export const verify = async (req, jwtPayload, done) => {

  if (jwtPayload) {
    if (!isAuthorized(jwtPayload)) {
      // tslint:disable-next-line:no-shadowed-variable
      const err = new Error('You do not have the proper role for signing');
      // err.code = 401;

      return done(err, null);
    }

    try {
      let userProfile;
      const dm = new DataManager(shared.pgPool);
      const { UserProfileModel } = dm;

      userProfile = await UserProfileModel.findByKeycloakId(jwtPayload.sub);
      if (!userProfile) {
        // create one
        userProfile = await UserProfileModel.create({
          keycloakId: jwtPayload.sub,
        });
      }

      const user = {
        roles: getJwtPayloadRoles(jwtPayload),
        name: jwtPayload.name,
        preferredUsername: jwtPayload.preferred_username,
        givenName: jwtPayload.given_name,
        familyName: jwtPayload.family_name,
        email: jwtPayload.email,
      };

      const profile: AuthenticatedUser = { ...userProfile, ...user };

      // The returned user will be made available via `req.user`
      return done(null, profile); // OK
    } catch (err) {
      const message = 'JWT verification error';
      logger.error(`${message}, err = ${err.message}`);
      return done(null, null); // FAIL
    }
  }

  // tslint:disable-next-line:no-shadowed-variable
  const err = new Error('Unable to authenticate');
  // err.code = 401;

  return done(err, false);
};

export const authmware = async app => {
  // app.use(session(sessionOptions));
  app.use(passport.initialize());
  app.use(passport.session());

  // We don't store any user information.
  passport.serializeUser((user, done) => {
    logger.info('serialize');
    done(null, {});
  });

  // We don't load any addtional user information.
  passport.deserializeUser((id, done) => {
    logger.info('deserialize');
    done(null, {});
  });

  const { certificate, algorithm } = await getJwtCertificate(config.get('sso:certsUrl'));
  const opts: any = {};
  opts.jwtFromRequest = ExtractJwt.fromAuthHeaderAsBearerToken();
  opts.algorithms = [algorithm];
  opts.secretOrKey = certificate;
  opts.passReqToCallback = true;
  // For development purposes only ignore the expiration
  // time of tokens.
  if (config.get('environment') === 'development') {
    opts.ignoreExpiration = true;
  }

  const jwtStrategy = new JwtStrategy(opts, verify);

  passport.use(jwtStrategy);
};
