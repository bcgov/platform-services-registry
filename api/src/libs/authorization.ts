//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { errorWithCode } from '@bcgov/common-nodejs-utils';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import shared from '../libs/shared';

const dm = new DataManager(shared.pgPool);

export const getAuthorization = async(profileId: number, user: any, projectDetails?:any, projectContacts?:any): Promise<boolean | Error> => {
  const { ProfileModel, ContactModel } = dm;

  if (!projectDetails){
    projectDetails = await ProfileModel.findById(Number(profileId));
  }

  if (!projectContacts){
    projectContacts = await ContactModel.findForProject(Number(profileId));
  }

  const authorizedEmails = projectContacts.map(contact => contact.email);

  if (!(
    user.id === projectDetails.userId
    || authorizedEmails.includes(user.email)
    || user.roles.includes(USER_ROLES.ADMINISTRATOR)
  )) {
    return errorWithCode('Unauthorized Access', 401);
  }

  return true;
}
