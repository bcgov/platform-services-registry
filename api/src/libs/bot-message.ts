//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import DataManager from '../db';
import { BotMessage } from '../db/model/request';
import { NatsContext } from '../types';
import shared from './shared';

const dm = new DataManager(shared.pgPool);
const { RequestModel, NamespaceModel } = dm;

export const createBotMessageSet = async ( requestId: number, natsSubject: string, natsContext: NatsContext ): Promise<any> => {
  if (!requestId) {
      throw new Error('Cant get profile id');
  }

  const clusters = await NamespaceModel.findClustersForProfile(natsContext.profile_id);

  for (const cluster of clusters) {
    await RequestModel.createBotMessage({
      requestId,
      natsSubject,
      natsContext,
      clusterName: cluster.name,
      receivedCallback: false,
    })
  }
};

export const fetchBotMessageRequests = async (requestId: number): Promise<BotMessage[]> => {
  try {
      return await RequestModel.findActiveBotMessagesByRequestId(requestId);
  } catch (err) {
      const message = `Unable to fetch existing bot message requests for request ${requestId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
  }
};
