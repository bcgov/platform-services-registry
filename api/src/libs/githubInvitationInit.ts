import { createAppAuth } from '@octokit/auth'
import { request } from '@octokit/request'
import { getGithubPrivateKey } from '../config/githubConfig'
import { logger } from '@bcgov/common-nodejs-utils';

// cached value
type APPInitailValue = {
  initialized: number | null,
  apps: object,
  nonInstallatedApp: Function | null,
}

const  installationApps = <APPInitailValue>{
  initialized: null,
  apps: {},
  nonInstallatedApp: null
}

/**
 * getNonInstallationApp
 * @returns a non installed github app
 */
export const getNonInstallationApp = () => {
    logger.info('getNonInstallationApp')
  // caches a non installed app
  try{


  if (!installationApps.nonInstallatedApp) {
    const auth = createAppAuth({
      appId: 128566,
      privateKey: getGithubPrivateKey(),
      clientId: process.env.CLIENT_ID,
      clientSecret: process.env.CLIENT_SECRET,
    })

    installationApps.nonInstallatedApp = request.defaults({
      request: {
        hook: auth.hook,
      },
      mediaType: {
        previews: ['machine-man'],
      },
    })
  }
 } catch(e){
    logger.error('I catch an error', e);
    

  }


  return installationApps.nonInstallatedApp
}

const newAuthorizedApp = (installationId) => {
  const app = createAppAuth({
    appId: 128566,
    privateKey: getGithubPrivateKey(),
    clientId: process.env.CLIENT_ID,
    clientSecret: process.env.CLIENT_SECRET,
    installationId,
  })

  return {
    initialized: Date.now(),
    app,
    id: installationId,
    authenticatedRequest: request.defaults({
      request: {
        hook: app.hook,
      },
      mediaType: {
        previews: ['machine-man'],
      },
    }),
  }
}

export const getInstallations = async () => {
    logger.info('getInstallations')

    const nonInstallationRequest = getNonInstallationApp()
   
    
    if(nonInstallationRequest !== null){
      const response = await nonInstallationRequest('GET /app/installations')
      return response.data
    }
    //TODO: maybe throw if there's no installation
    return []
}

export const getOrgInstallations = async () => {
  logger.info('getOrgInstallations')

  const installations = await getInstallations()

  // lower case installation login names
  const loweredInstallations = installations.map((i) => ({
    ...i,
    account: { ...i.account, login: i.account.login.toLowerCase() },
  }))

  logger.info(
    `This github app has been installed on ${loweredInstallations.map(
      (i) => i.account.login
    )}`
  )

  return installations
}

/**
 * a new authenticated app must be created for every installation in order to invite users
 */
export const getAuthenticatedApps =  async () => {

  try {
    logger.info('getAuthenticatedApps')
    if (!installationApps.initialized) {
      logger.info('Initializing Authenticated Apps')
      installationApps.initialized = Date.now()
      
      const installations = await getOrgInstallations()
      
      
      installations.forEach((installation) => {
 
        const name = installation.account.login.toLowerCase()
     
        if (!installationApps.apps[name]) {
          logger.info(
            `newAuthorizedApp created for ${name} installation: ${installation.id}`
            )
            installationApps.apps[name] = newAuthorizedApp(installation.id)
          }
        })
      } else {
        logger.info(
          `Installation Apps returned: ${Object.keys(installationApps.apps)}`
          )
          logger.info(
            `Authenticated Apps were cached, reusing the ones initialized on ${installationApps.initialized}`
            )
          }
        } catch(e){
         console.error(e)
        }

  return installationApps
}

/**
 * initializes and validates SSO config as well as github applications
 * all errors bubble to top to quit process
 */
export const init = async () => {
    logger.info('Checking Authenticated Apps')
  await getAuthenticatedApps()
}


export default init
