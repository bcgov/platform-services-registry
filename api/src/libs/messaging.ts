//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { logger } from '@bcgov/common-nodejs-utils';
import fs from 'fs';
import path from 'path';
import config from '../config';
import DataManager from '../db';
import { ProjectProfile } from '../db/model/profile';
import { HumanAction } from '../db/model/request';
import { BodyType, Message, SendReceipt } from '../libs/service';
import shared from './shared';
import { transformContacts } from './utils';

export const enum MessageType {
  ProvisioningStarted = 0,
  ProvisioningCompleted,
  EditRequestStarted,
  EditRequestCompleted,
  RequestApproval,
  RequestRejected,
}

const dm = new DataManager(shared.pgPool);
const { ProfileModel, ContactModel, RequestModel} = dm;

const updateEmailContent = async (
    buff: string, profile: ProjectProfile, contactDetails: any, request: any, humanAction: any
  ): Promise<string> => {
  try {
    let emailContent: string;
    let humanActionComment: string = '';

    if (humanAction) {
      humanActionComment = (humanAction.comment !== null ? `Additional information: ${humanAction.comment}` : '');
    }

    const mapObj = {
      POName: contactDetails.POName,
      TCName: contactDetails.TCName,
      POEmail: contactDetails.POEmail,
      TCEmail: contactDetails.TCEmail,
      POGitHub: contactDetails.POGithubId,
      TCGitHub: contactDetails.TCGithubId,
      projectName: profile.name,
      projectDescription : profile.description,
      projectMinistry: profile.busOrgId,
      setCluster: profile.primaryClusterName,
      licensePlate: `${profile.namespacePrefix}`,
      requestType: ( request.editType ? 'Project Edit': 'New Project'),
      quotaSize: ( request.editType === 'quotaSize' ? request.editObject.quota : 'Small'),
      editType: ( request.editType === 'quotaSize' ? 'Quota' : ''),
      humanActionComment,
    };

    const re = new RegExp(Object.keys(mapObj).join('|'),'gi');
    emailContent = buff.replace(re, matched => {
      return mapObj[matched];
    });
    return emailContent;
  } catch (err) {
    const message = `Unable to update email content`;
    logger.error(`${message}, err = ${err.message}`);

    return '';
  }

}

export const sendProvisioningMessage = async (profileId: number, messageType: MessageType): Promise<SendReceipt | undefined> => {

  try {
    const profile = await ProfileModel.findById(profileId);

    const reviewerEmails = config.get('reviewers:emails');
    const contacts = await ContactModel.findForProject(profileId);
    const contactDetails = await transformContacts(contacts)
    const to = (
      messageType === MessageType.RequestApproval ? reviewerEmails : [...new Set(contacts.map(c => c.email))]
      );
    const requests = await RequestModel.findForProfile(profileId);
    const request = requests.pop();
    if (!request) {
      return
    }

    const humanAction: HumanAction | undefined = await RequestModel.findHumanActionByRequestId(Number(request.id));

    let buff;

    if (to.length === 0) {
      return;
    }

    switch (messageType) {
      case MessageType.ProvisioningStarted:
        buff = fs.readFileSync(
          path.join(__dirname, '../../', 'templates/provisioning-request-received.html')
          ).toString();
        break;
      case MessageType.ProvisioningCompleted:
        buff = fs.readFileSync(
          path.join(__dirname, '../../', 'templates/provisioning-request-done.html')
          ).toString();
        break;
      case MessageType.EditRequestStarted:
        buff = fs.readFileSync(
          path.join(__dirname, '../../', 'templates/edit-request-received.html')
          ).toString();
        break;
      case MessageType.EditRequestCompleted:
        buff = fs.readFileSync(
          path.join(__dirname, '../../', 'templates/edit-request-done.html')
          ).toString();
        break;
      case MessageType.RequestApproval:
        buff = fs.readFileSync(
          path.join(__dirname, '../../', 'templates/request-approval.html')
          ).toString();
        break;
      case MessageType.RequestRejected:
        buff = fs.readFileSync(
          path.join(__dirname, '../../', 'templates/request-rejected.html')
          ).toString();
        break;
      default:
        logger.info('No message type given');
        return;
    }

    if (!buff) {
      return;
    }

    const bodyContent = await updateEmailContent(buff, profile, contactDetails, request, humanAction);

    if (!bodyContent) {
      return;
    }

    const message: Message = {
      bodyType: BodyType.HTML,
      body: bodyContent,
      to,
      from: 'Registry <PlatformServicesTeam@gov.bc.ca>',
      subject: `${profile.name} OCP 4 Project Set`,
    }

    const receipt = await shared.ches.send(message);
    logger.info(`Message (${messageType}) sent with transaction details: ${JSON.stringify(receipt)}`);

    return receipt;
  } catch (err) {
    const message = `Unable to send message for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    return;
  }
}
