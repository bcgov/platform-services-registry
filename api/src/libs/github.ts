import { logger } from '@bcgov/common-nodejs-utils';
import { getConfig } from '../config/githubConfig';
import { getAuthenticatedApps } from './githubInvitationInit';


export const inviteUserToOrg = async (userId, org) => {
  const installations = await getAuthenticatedApps()
  const app = installations.apps[org.toLowerCase()]
  return app.authenticatedRequest('POST /orgs/{org}/invitations', {
    org,
    invitee_id: userId,
  })
}

export const inviteUserToOrgs = async (userId, orgs) => {
  logger.info(`inviteUserToOrg ${userId} ${orgs}`)
  return orgs.map(async (org) => {
    try {
      logger.info(`User ${userId} is being invited to ${org}`)
      await inviteUserToOrg(userId, org)
      logger.info(`User ${userId} is to ${org}`)
    } catch (err:any) {
      if (err.status === 422) {
        logger.info(`User ${userId} is already in org ${org}. Skipping.`)
      } else {
        logger.error(`Could not invite ${userId} to org ${org}`)
        logger.info(JSON.stringify(err))
        throw err
      }
    }
  })
}

export const getUserByName = async (username) => {
  logger.info(`getUserByName ${username}`)
  const installations = await getAuthenticatedApps()
  const primaryOrg = getConfig().primaryOrg.toLowerCase()

  const response = await installations.apps[
    primaryOrg
  ].authenticatedRequest('GET /users/{username}', {
    username,
  })
  return response.data
}
