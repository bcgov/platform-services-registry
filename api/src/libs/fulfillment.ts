//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import config from '../config';
import { ROLE_IDS } from '../constants';
import DataManager from '../db';
import { Contact } from '../db/model/contact';
import { ProjectProfile } from '../db/model/profile';
import { Quotas, QuotaSize } from '../db/model/quota';
import { RequestEditType } from '../db/model/request';
import { replaceForDescription } from '../libs/utils';
import { NatsContext, NatsContextAction, NatsContextType, NatsMessage } from '../types';
import { MessageType, sendProvisioningMessage } from './messaging';
import { getQuotaSize } from './profile';
import shared from './shared';

const dm = new DataManager(shared.pgPool);
const { ProfileModel, ContactModel, QuotaModel, NamespaceModel } = dm;

export const fulfillNamespaceProvisioning = async (profileId: number) =>
  new Promise(async (resolve, reject) => {
    try {
      const profile: ProjectProfile = await ProfileModel.findById(profileId);
      const subjectPrefix: string = config.get('nats:subjectPrefix');

      const subject: string = subjectPrefix.concat(profile.primaryClusterName);
      const context: NatsContext = await contextForProvisioning(profileId, false);

      await sendNatsMessage(profileId, {
        natsSubject: subject,
        natsContext: context,
      });

      logger.info(`Sending CHES message (${MessageType.ProvisioningStarted}) for ${profileId}`);
      await sendProvisioningMessage(profileId, MessageType.ProvisioningStarted);
      logger.info(`CHES message sent for ${profileId}`);

      resolve();
    } catch (err) {
      const message = `Unable to fulfill namespace provisioning for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  });

export const fulfillEditRequest = async (profileId: number, requestEditType: RequestEditType, requestEditObject: any): Promise<NatsMessage> =>
  new Promise(async (resolve, reject) => {
    try {
      const profile: ProjectProfile = await ProfileModel.findById(profileId);
      const subjectPrefix: string = config.get('nats:subjectPrefix');

      const subject: string = subjectPrefix.concat(profile.primaryClusterName);
      const context: NatsContext = await contextForEditing(profileId, false, requestEditType, requestEditObject);

      const natsMessage = await sendNatsMessage(profileId, {
        natsSubject: subject,
        natsContext: context,
      });

      resolve(natsMessage);
    } catch (err) {
      const message = `Unable to fulfill namespace provisioning for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  });

export const contextForProvisioning = async (profileId: number, isForSync: boolean): Promise<NatsContext> => {
  try {
    const action = isForSync ? NatsContextAction.Sync : NatsContextAction.Create;
    const profile: ProjectProfile = await ProfileModel.findById(profileId);
    const contacts: Contact[] = await ContactModel.findForProject(profileId);
    const quotaSize: QuotaSize = await getQuotaSize(profile);
    const quotas: Quotas = await QuotaModel.findForQuotaSize(quotaSize);

    return await buildContext(action, profile, contacts, quotaSize, quotas);
  } catch (err) {
    const message = `Unable to create context for provisioning ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

export const contextForEditing = async (profileId: number, isForSync: boolean, requestEditType: RequestEditType, requestEditObject: any): Promise<NatsContext> => {
  try {
    const action = isForSync ? NatsContextAction.Sync : NatsContextAction.Edit;
    let profile: ProjectProfile;
    let quotaSize: QuotaSize;
    let quotas: Quotas;
    let contacts: Contact[];

    if (requestEditType === RequestEditType.ProjectProfile) {
      profile = requestEditObject;
    } else {
      profile = await ProfileModel.findById(profileId);
    }

    if (requestEditType === RequestEditType.QuotaSize) {
      quotaSize = requestEditObject.quota;
      quotas = requestEditObject.quotas;
    } else {
      quotaSize = await getQuotaSize(profile);
      quotas = await QuotaModel.findForQuotaSize(quotaSize);
    }

    if (requestEditType === RequestEditType.Contacts) {
      contacts = requestEditObject;
    } else {
      contacts = await ContactModel.findForProject(profileId);
    }

    return await buildContext(action, profile, contacts, quotaSize, quotas);
  } catch (err) {
    const message = `Unable to create context for updating ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

const buildContext = async (
  action: NatsContextAction, profile: ProjectProfile, contacts: Contact[], quotaSize: QuotaSize, quotas: Quotas
): Promise<NatsContext> => {
  try {
    if (!profile.id) {
      throw new Error('Cant get profile id');
    }
    const namespaces = await NamespaceModel.findForProfile(profile.id);

    const tcContact = contacts.filter(c => c.roleId === ROLE_IDS.TECHNICAL_CONTACT).pop();
    const poContact = contacts.filter(c => c.roleId === ROLE_IDS.PRODUCT_OWNER).pop();

    if (!profile || !tcContact || !poContact || !quotaSize || !quotas || !namespaces) {
      throw new Error('Missing arguments to build nats context');
    }

    return {
      action,
      type: NatsContextType.Standard,
      profileId: profile.id,
      displayName: profile.name,
      newDisplayName: 'NULL',
      description: profile.description,
      quota: quotaSize,
      quotas,
      namespaces,
      technicalContact: {
        userId: tcContact.githubId,
        provider: 'github', // TODO:(JL) Fix as part of #94.
        email: tcContact.email,
      },
      productOwner: {
        userId: poContact.githubId,
        provider: 'github', // TODO:(JL) Fix as part of #94.
        email: poContact.email,
      },
    };
  } catch (err) {
    const message = `Unable to build context for profile ${profile.id}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};

const sendNatsMessage = async (profileId: number, natsMessage: NatsMessage): Promise<NatsMessage> => {
  try {
    const nc = shared.nats;
    const { natsSubject, natsContext } = natsMessage;

    nc.on('error', () => {
      const errmsg = `NATS error sending order ${profileId} to ${natsSubject}`;
      throw new Error(errmsg);
    });

    logger.info(`Sending NATS message for ${profileId} to ${natsSubject}`);

    nc.publish(natsSubject, replaceForDescription(natsContext));
    logger.info(`NATS Message sent for ${profileId} to ${natsSubject}`);

    nc.flush(() => {
      nc.removeAllListeners(['error']);
    });

    return natsMessage;
  } catch (err) {
    const message = `Unable to send nats message for profile ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw err;
  }
};
