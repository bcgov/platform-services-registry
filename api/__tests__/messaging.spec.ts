//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import mockAxios from 'axios';
import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { contactsForProfile, MessageType, sendProvisioningMessage } from '../src/libs/messaging';

const p1 = path.join(__dirname, 'fixtures/select-profile-contacts.json');
const contacts = JSON.parse(fs.readFileSync(p1, 'utf8'));

const p2 = path.join(__dirname, 'fixtures/post-send-email-resp.json');
const send = JSON.parse(fs.readFileSync(p2, 'utf8'));


describe('Services', () => {

  const client = new Pool().connect();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Contact details are returned', async () => {

    client.query.mockReturnValue({ rows: contacts });

    const result = await contactsForProfile(12345);

    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
    expect(client.query.mock.calls).toMatchSnapshot();
  });

  it('Contact details not returned', async () => {

    client.query.mockImplementation(() => { throw new Error() });

    const result = await contactsForProfile(12345);

    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
    expect(client.query.mock.calls).toMatchSnapshot();
  });

  it('Provisioning started message is sent', async () => {
    // @ts-ignore
    mockAxios.fn.post.mockImplementationOnce(() =>
      Promise.resolve({
        data: send,
      })
    );
    client.query.mockReturnValue({ rows: contacts });

    const result = await sendProvisioningMessage(12345, MessageType.ProvisioningStarted);

    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
  });

  it('Provisioning completed message is sent', async () => {
    // @ts-ignore
    mockAxios.fn.post.mockImplementationOnce(() =>
      Promise.resolve({
        data: send,
      })
    );
    client.query.mockReturnValue({ rows: contacts });

    const result = await sendProvisioningMessage(12345, MessageType.ProvisioningCompleted);

    expect(result).toBeDefined();
    expect(result).toMatchSnapshot();
  });

  it('No message type results in a message not being sent', async () => {
    // @ts-ignore
    mockAxios.fn.post.mockImplementationOnce(() =>
      Promise.resolve({
        data: send,
      })
    );
    client.query.mockReturnValue({ rows: contacts });

    const result = await sendProvisioningMessage(12345, 555);

    expect(result).not.toBeDefined();
  });

  it('No contacts results in a message not being sent', async () => {
    // @ts-ignore
    mockAxios.fn.post.mockImplementationOnce(() =>
      Promise.resolve({
        data: send,
      })
    );
    client.query.mockReturnValue({ rows: [] });

    const result = await sendProvisioningMessage(12345, MessageType.ProvisioningStarted);

    expect(result).not.toBeDefined();
  });
});
