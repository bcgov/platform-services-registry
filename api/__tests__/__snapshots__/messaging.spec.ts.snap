// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services Contact details are returned 1`] = `
Array [
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.436Z",
    "email": "jane@example.com",
    "firstName": "Jane",
    "githubId": "jane1100",
    "id": 1,
    "lastName": "Doe",
    "roleId": 1,
    "updatedAt": "2020-09-10T18:14:13.436Z",
  },
  Object {
    "archived": false,
    "createdAt": "2020-09-10T18:14:13.436Z",
    "email": "jim@example.com",
    "firstName": "Jim",
    "githubId": "jim1100",
    "id": 2,
    "lastName": "Doe",
    "roleId": 2,
    "updatedAt": "2020-09-10T18:14:13.436Z",
  },
]
`;

exports[`Services Contact details are returned 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT contact.*
          FROM contact
          JOIN profile_contact ON contact.id = profile_contact.contact_id
            WHERE profile_contact.profile_id = $1;
      ",
      "values": Array [
        12345,
      ],
    },
  ],
]
`;

exports[`Services Contact details not returned 1`] = `Array []`;

exports[`Services Contact details not returned 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT contact.*
          FROM contact
          JOIN profile_contact ON contact.id = profile_contact.contact_id
            WHERE profile_contact.profile_id = $1;
      ",
      "values": Array [
        12345,
      ],
    },
  ],
]
`;

exports[`Services Provisioning completed message is sent 1`] = `
Object {
  "messages": Array [
    Object {
      "messageId": "wcjrijro-6fca-4a78-ba16-fb93d094daa3",
      "to": Array [
        "phill@example.com",
      ],
    },
  ],
  "transactionId": "wcjrijro-5fd9-499b-8368-72b70196cf55",
}
`;

exports[`Services Provisioning started message is sent 1`] = `
Object {
  "messages": Array [
    Object {
      "messageId": "wcjrijro-6fca-4a78-ba16-fb93d094daa3",
      "to": Array [
        "phill@example.com",
      ],
    },
  ],
  "transactionId": "wcjrijro-5fd9-499b-8368-72b70196cf55",
}
`;
